File: runtimes/python/helix_rt/gateway.py
import time


class TenantLimiter:
    def __init__(self, capacity: float, rate: float):
        self.capacity = capacity
        self.rate = rate
        self.tokens = capacity
        self.last_refill = time.time()

    def consume(self, count: float) -> bool:
        now = time.time()
        elapsed = now - self.last_refill
        self.tokens = min(self.capacity, self.tokens + elapsed * self.rate)
        self.last_refill = now

        if self.tokens >= count:
            self.tokens -= count
            return True
        return False


class MultiTenantRateLimiter:
    def __init__(self, default_capacity: float, default_rate: float):
        self.default_capacity = default_capacity
        self.default_rate = default_rate
        self.limiters = {}

    def set_tenant_limit(self, tenant_id: str, capacity: float, rate: float) -> None:
        self.limiters[tenant_id] = TenantLimiter(capacity, rate)

    def allow(self, tenant_id: str, count: float) -> bool:
        limiter = self.limiters.get(tenant_id)
        if not limiter:
            limiter = TenantLimiter(self.default_capacity, self.default_rate)
            self.limiters[tenant_id] = limiter
        return limiter.consume(count)

    def allow_jwt_request(self, token: str, secret: str, count: float) -> bool:
        try:
            import jwt
            payload = jwt.decode(token, secret, algorithms=["HS256"])
            tenant_id = payload.get("tenant_id")
            if not tenant_id:
                return False
            return self.allow(tenant_id, count)
        except Exception:
            return False

File: runtimes/java/src/main/java/io/helixrpc/runtime/Gateway.java
package io.helixrpc.runtime;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class Gateway {
    private static class TenantLimiter {
        private final double capacity;
        private final double rate;
        private double tokens;
        private long lastRefill;

        public TenantLimiter(double capacity, double rate) {
            this.capacity = capacity;
            this.rate = rate;
            this.tokens = capacity;
            this.lastRefill = System.nanoTime();
        }

        public synchronized boolean consume(double count) {
            long now = System.nanoTime();
            double elapsed = (now - lastRefill) / 1e9;
            tokens = Math.min(capacity, tokens + elapsed * rate);
            lastRefill = now;

            if (tokens >= count) {
                tokens -= count;
                return true;
            }
            return false;
        }
    }

    public static class MultiTenantRateLimiter {
        private final Map<String, TenantLimiter> limiters = new ConcurrentHashMap<>();
        private final double defaultCapacity;
        private final double defaultRate;

        public MultiTenantRateLimiter(double defaultCapacity, double defaultRate) {
            this.defaultCapacity = defaultCapacity;
            this.defaultRate = defaultRate;
        }

        public void setTenantLimit(String tenantId, double capacity, double rate) {
            limiters.put(tenantId, new TenantLimiter(capacity, rate));
        }

        public boolean allow(String tenantId, double count) {
            TenantLimiter limiter = limiters.computeIfAbsent(tenantId, 
                id -> new TenantLimiter(defaultCapacity, defaultRate));
            return limiter.consume(count);
        }
    }
}
